text/0021-any-and-all-operators.md

```diff
@@ -0,0 +1,165 @@
+
+For example, consider the expression `[1, true].all(it like "foo")`. Regardless of the order in which the predicate is applied, the expression returns the same `QuantifierError`.
+
+This semantics lets us think about `all` and `any` in terms of their expansion to conjunctions and disjunctions. The quantified expression errors if and only if the expansion would error.
+
+For example:
+```
+// Throws QuantifierError
+[1, true].all(it like "foo")
+
+// Throws a TypeError: "like" can't be applied to an integer
+1 like "foo" && true like "foo"
+
+// Throws a TypeError: "like" can't be applied to an boolean
+true like "foo" && 1 like "foo"
+```
+
+The only difference is the error being thrown: `QuantifierError` for the quantified expression versus an element-specific error for the explicit expansion.
+
+### Extending the type system
+
+The type system needs to be extended with rules for type checking quantified expressions `E.all(P)` and `E.any(P)`. This should be fairly standard: we know the type of `it` from the type of the set `E`, and we can use this to typecheck the predicate `P`. We also need to propagate effects through predicates.
+
+For example, the following should typecheck when `context.limit` is an optional integer attribute, and `resource.things` is a set of entities with an optional integer attribute `.size`:
+
+```
+context has limit &&
+resource.things.all(
+  it has size && it.size < context.limit
+)
+```
+
+## Drawbacks
+
+- This proposal requires significant development effort. We'll need to modify the parser, the CST to AST conversion, CST/AST/EST, evaluator, validator, models, proofs, and DRT.
+- The `any` and `all` operators may encourage writing complex expressions with poor performance. For example, `context.portNumbers.any(it == 8010)` is $O(n)$ versus $O(1)$ for `context.portNumbers.contains(8010)`.
+- Supporting these operators will likely have a negative impact on analysis performance, especially for complex predicates.
+
+## Alternatives
+
+We considered two alternative semantics for these operators.
+
+### Non-deterministic error behavior
+
+Given `E.any(P)` or `E.all(P)`, we apply `P` to elements of `E` in arbitrary order, terminating on the first error and returning that error as the result.
+
+Under this semantics, the same policy could produce different errors depending on iteration order. For example, `[1, true].all(it like "foo")` may result in a type error that says `like` cannot be applied to an integer or a boolean. In other words, Cedar semantics would become non-determinstic.
+
+We decided against this semantics for three reasons:
+* Violates Cedar's design as a deterministic language. Error non-determinism would be visible to `is_authorized` callers.
+* Applications may also become non-deterministic if they examine errors and act on them.
+* Non-determinism complicates models and proofs.
+
+Keeping deterministic semantics is important for Cedar as an authorization language.
+
+### Treating errors as `false`
```

Comment:
A big advantage of this approach is that it is much cheaper to analyze than any of the alternatives that propagate errors.  Will add to the RFC.